### PR TITLE
feat: [M3-7948] - Add Region filtering to Linodes landing table

### PR DESCRIPTION
**Pull request by hana-linode**
## Description 📝
Add support for region filtering in the Linodes landing table for Gecko LA. Region filter is saved in local storage similar to the "Show X" dropdown.

## Changes  🔄
List any change relevant to the reviewer.
- Add `site_type` to the linode instance type
- Removed `edge` references where possible
- Add region filter dropdown to the Linodes landing table

## Preview 📷
(ip address has been redacted)

https://github.com/user-attachments/assets/c538c7e2-6c46-415b-aedd-66683516cded

### Verification steps
- `All` should display all Linodes (core and distributed regions)
- `Core` should only display Linodes in `core` regions
- `Distributed` should only display Linodes in `distributed` regions
- Ensure sorting, summary, and tag view still work as normal
- Navigate to another page and go back to the Landing page, the region selection should stay the same

## As an Author I have considered 🤔

*Check all that apply*

- [x] 👀 Doing a self review
- [ ] ❔ Our [contribution guidelines](https://github.com/linode/manager/blob/develop/docs/CONTRIBUTING.md)
- [x] 🤏 Splitting feature into small PRs
- [x] ➕ Adding a changeset
- [ ] 🧪 Providing/Improving test coverage
- [ ] 🔐 Removing all sensitive information from the code and PR description
- [x] 🚩 Using a feature flag to protect the release
- [x] 👣 Providing comprehensive reproduction steps
- [ ] 📑 Providing or updating our documentation
- [ ] 🕛 Scheduling a pair reviewing session
- [ ] 📱 Providing mobile support
- [ ] ♿  Providing accessibility support

### CODE REVIEW

File: packages/api-v4/src/linodes/types.ts [hunk 1]
```diff
@@ -36,6 +36,7 @@ export interface Linode {
   specs: LinodeSpecs;
   watchdog_enabled: boolean;
   tags: string[];
+  site_type: RegionSite;
```
Comment (mjac0bs) [inline packages/api-v4/src/linodes/types.ts:39]: Can we get a changeset for this apiv4 change?

File: packages/manager/src/features/Linodes/LinodesLanding/DisplayLinodes.tsx [hunk 2]
```diff
@@ -132,17 +163,48 @@ export const DisplayLinodes = React.memo((props: DisplayLinodesProps) => {
...
+                            handleRegionFilter(selectedOption.value);
+                          }
+                        }}
+                        sx={{
+                          display: 'inline-block',
+                        }}
+                        textFieldProps={{
+                          hideLabel: true,
+                        }}
+                        disableClearable
```
Comment (mjac0bs) [inline packages/manager/src/features/Linodes/LinodesLanding/DisplayLinodes.tsx:189]: If this value has never been set, or if local storage is cleared, we end up with placeholder text that is cut off due to the width of the autocomplete field.

Can we either increase the width or default to `All` if there isn't a selection in local storage?

Screenshot 2024-07-17 at 2 11 30 PM
Screenshot 2024-07-17 at 2 12 03 PM

File: packages/manager/src/features/Linodes/LinodesLanding/ListView.tsx
Comment (mjac0bs) [inline packages/manager/src/features/Linodes/LinodesLanding/ListView.tsx:1]: Just to confirm that we didn't unintentionally overlook it: the filtering on region type is only for List View, and we have no plans to include it in Card View?

https://github.com/user-attachments/assets/835ce8ab-8a0d-4263-9105-e4af60d07498

**Review by mjac0bs — COMMENTED** [review]
x-filter in mock requests looks good and so does the persistence of the selection.

File: packages/manager/src/features/Linodes/LinodesLanding/DisplayLinodes.tsx (hunk 2 above)
Comment (hana-linode) [reply packages/manager/src/features/Linodes/LinodesLanding/DisplayLinodes.tsx:189]: Yep, defaulted to `All`

File: packages/manager/src/features/Linodes/LinodesLanding/ListView.tsx
Comment (hana-linode) [reply packages/manager/src/features/Linodes/LinodesLanding/ListView.tsx:1]: Woops, good catch! That was unintentional, just added region filtering to card view as well

**Review by mjac0bs — COMMENTED** [review]
Thanks for the changes! `All` as a default is great, card view looks good, and putting the region type filter in its own component is clean.

Let's get the merge conflict resolved and make sure CI passes. And I think we do actually need to extend the width of the autocomplete after all. The checkmark is too close to the selection: 

Screenshot 2024-07-18 at 1 39 18 PM

File: packages/api-v4/.changeset/pr-10639-added-1721317037923.md [hunk 3]
```diff
@@ -0,0 +1,5 @@
+---
+"@linode/api-v4": Added
+---
+
+site_type to the linode instance type ([#10639](https://github.com/linode/manager/pull/10639))
```
Comment (carrillo-erik) [inline packages/api-v4/.changeset/pr-10639-added-1721317037923.md:5]: Include 'Add' at the beginning of the sentence.

File: packages/manager/src/features/Linodes/LinodesLanding/RegionTypeFilter.tsx [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+interface Props {
+  handleRegionFilter: (regionFilter: RegionFilter) => void;
+}
+
+export const RegionTypeFilter = (props: Props) => {
+  const { handleRegionFilter } = props;
+
+  return (
+    <label style={{ alignItems: 'center', display: 'flex' }}>
+      <span style={{ marginLeft: 8, marginRight: 8 }}>Region Type:</span>{' '}
```
Comment (carrillo-erik) [inline packages/manager/src/features/Linodes/LinodesLanding/RegionTypeFilter.tsx:37]: Could these be replaced with MUI components instead of using native html?

**Comment by carrillo-erik** [comment]
I tested the functionality as described and the changes work as expected. I left two comments which are non-blocking items. Just need to ensure the tests pass before merging.

**Review by carrillo-erik — APPROVED** [review]

File: packages/api-v4/.changeset/pr-10639-added-1721317037923.md (hunk 3 above)
Comment (bnussman-akamai) [reply packages/api-v4/.changeset/pr-10639-added-1721317037923.md:5]: > Avoid starting a changeset with the verb "Add", "Remove", "Change" or "Fix" when listed under that respective Added, Removed, Changed or Fixed section. It is unnecessary repetition.

Changeset is correct according to [our docs](https://linode.github.io/manager/CONTRIBUTING.html#writing-a-changeset). If anything we could put `site_type` inside of a code snippet

```suggestion
`site_type` to the linode instance type ([#10639](https://github.com/linode/manager/pull/10639))
```

File: packages/manager/src/features/Linodes/LinodesLanding/DisplayGroupedLinodes.tsx [hunk 5]
```diff
@@ -207,79 +223,88 @@ export const DisplayGroupedLinodes = (props: DisplayGroupedLinodesProps) => {
...
-                  pageSize,
-                }) => {
-                  const finalProps = {
-                    ...rest,
+      <>
+        {isGeckoLAEnabled && (
+          <Paper sx={{ padding: 1 }}>
+            <RegionTypeFilter handleRegionFilter={handleRegionFilter} />
+          </Paper>
+        )}
```
Comment (hana-linode) [inline packages/manager/src/features/Linodes/LinodesLanding/DisplayGroupedLinodes.tsx:231]: This was the main change

File: packages/manager/src/features/Linodes/LinodesLanding/LinodesLanding.tsx [hunk 6]
```diff
@@ -241,7 +247,7 @@ class ListLinodes extends React.Component<CombinedProps, State> {
       return <CircleProgress />;
     }
 
-    if (this.props.linodesData.length === 0) {
+    if (totalNumLinodes === 0 && linodesData.length === 0) {
```
Comment (hana-linode) [inline packages/manager/src/features/Linodes/LinodesLanding/LinodesLanding.tsx:250]: We can't just rely on the length of `linodesData` because the `0` could be coming from filtered data and the user could have still have linodes

File: packages/api-v4/.changeset/pr-10639-added-1721317037923.md
Comment (mjac0bs) [inline packages/api-v4/.changeset/pr-10639-added-1721317037923.md:1]: Was this done in another PR?

File: packages/manager/src/features/Linodes/LinodesLanding/DisplayGroupedLinodes.tsx [hunk 7]
```diff
@@ -93,44 +100,53 @@ export const DisplayGroupedLinodes = (props: DisplayGroupedLinodesProps) => {
...
-            </Tooltip>
+              <div className="visually-hidden" id={groupByDescriptionId}>
+                {linodesAreGrouped
+                  ? 'group by tag is currently enabled'
+                  : 'group by tag is currently disabled'}
+              </div>
+              <Tooltip placement="top-end" title="Ungroup by tag">
+                <StyledToggleButton
+                  aria-describedby={groupByDescriptionId}
+                  aria-label={`Toggle group by tag`}
```
Comment (mjac0bs) [inline packages/manager/src/features/Linodes/LinodesLanding/DisplayGroupedLinodes.tsx:140]:
```suggestion
                  aria-label="Toggle group by tag"
```
Nit

File: packages/manager/src/features/Linodes/LinodesLanding/DisplayLinodes.tsx [hunk 8]
```diff
@@ -132,58 +143,76 @@ export const DisplayLinodes = React.memo((props: DisplayLinodesProps) => {
...
-                      </StyledToggleButton>
-                    </Tooltip>
+                        {linodesAreGrouped
+                          ? 'group by tag is currently enabled'
+                          : 'group by tag is currently disabled'}
+                      </div>
+                      <Tooltip placement="top-end" title="Group by tag">
+                        <StyledToggleButton
+                          aria-describedby={groupByDescriptionId}
+                          aria-label={`Toggle group by tag`}
```
Comment (mjac0bs) [inline packages/manager/src/features/Linodes/LinodesLanding/DisplayLinodes.tsx:206]:
```suggestion
                          aria-label="Toggle group by tag"
```
Same nit

**Review by mjac0bs — COMMENTED** [review]
UI: I was seeing a minor issue when one of the categories has no Linodes, Summary View is selected, and we are *not* grouping by tags. Can we add a consistent amount of spacing before the "No items to display"?



Code: Removal of `edge` looked good. Saw just a few nits.

https://github.com/user-attachments/assets/a1172061-b298-42bb-a626-31f7b8b900ae

File: packages/api-v4/.changeset/pr-10639-added-1721317037923.md
Comment (hana-linode) [reply packages/api-v4/.changeset/pr-10639-added-1721317037923.md:1]: Oops, good catch! 🔍  Yes, this was added before the other PR was merged so it's no longer necessary. Just removed

**Review by bnussman-akamai — APPROVED** [review]
Looks good! 

I wonder if we should add some kind of small loading state for when the user is switching between different region filters

**Review by mjac0bs — APPROVED** [review]
🚀

**Comment by hana-linode** [comment]
@bnussman-akamai Good point, just realized we don't use the `TableRowLoading` component in `LinodesLanding`. I'll create a separate ticket to address that since it's going to require some code maneuvering
